File: backend/app/schemas/bug.py
"""Pydantic schemas for bug detection."""

from typing import List, Dict, Any, Optional
from pydantic import BaseModel, Field


class BugEvidence(BaseModel):
    """Evidence for bug detection."""

    execution_plan: Optional[Dict[str, Any]] = Field(None, description="Plan evidence")
    parameters: Optional[Dict[str, Any]] = Field(None, description="Parameter evidence")
    query_characteristics: Optional[Dict[str, Any]] = Field(
        None, description="Query characteristic evidence"
    )
    wait_events: Optional[Dict[str, Any]] = Field(None, description="Wait event evidence")
    sql_characteristics: Optional[Dict[str, Any]] = Field(
        None, description="SQL pattern evidence"
    )
    alert_log: Optional[Dict[str, Any]] = Field(None, description="Alert log evidence")

    class Config:
        from_attributes = True


class Bug(BaseModel):
    """Oracle bug information."""

    bug_number: str = Field(..., description="Oracle bug number")
    title: str = Field(..., description="Bug title")
    category: str = Field(..., description="Bug category")
    severity: str = Field(..., description="Severity level")
    description: str = Field(..., description="Bug description")
    symptoms: List[str] = Field(..., description="Bug symptoms")
    affected_versions: List[str] = Field(..., description="Affected Oracle versions")
    fixed_versions: List[str] = Field(..., description="Fixed in versions")
    workarounds: List[str] = Field(..., description="Workaround suggestions")
    remediation_sql: Optional[str] = Field(None, description="SQL to remediate")
    my_oracle_support_url: Optional[str] = Field(None, description="MOS doc URL")

    class Config:
        from_attributes = True


class BugMatch(BaseModel):
    """Detected bug match."""

    bug: Bug = Field(..., description="Bug information")
    confidence: int = Field(..., description="Confidence score (0-100)")
    matched_patterns: List[str] = Field(..., description="Matched detection patterns")
    evidence: BugEvidence = Field(..., description="Evidence for detection")
    sql_id: Optional[str] = Field(None, description="Related SQL_ID")

    class Config:
        from_attributes = True


class BugDetectionSummary(BaseModel):
    """Summary of bug detection results."""

    total_bugs: int = Field(..., description="Total bugs detected")
    by_severity: Dict[str, int] = Field(..., description="Count by severity")
    by_category: Dict[str, int] = Field(..., description="Count by category")
    high_confidence_count: int = Field(
        ..., description="Number of high confidence matches"
    )
    critical_count: int = Field(..., description="Number of critical bugs")

    class Config:
        from_attributes = True


class BugDetectionResponse(BaseModel):
    """API response for bug detection."""

    sql_id: Optional[str] = Field(None, description="SQL_ID analyzed")
    detected_bugs: List[BugMatch] = Field(..., description="List of detected bugs")
    summary: BugDetectionSummary = Field(..., description="Detection summary")
    database_version: Optional[str] = Field(None, description="Database version")
    detection_timestamp: str = Field(..., description="When detection was performed")


class BugListResponse(BaseModel):
    """API response for bug list."""

    bugs: List[Bug] = Field(..., description="List of bugs")
    total_count: int = Field(..., description="Total number of bugs")
    filters_applied: Optional[Dict[str, Any]] = Field(
        None, description="Applied filters"
    )


class AlertLogBugResponse(BaseModel):
    """API response for alert log bug detection."""

    detected_bugs: List[BugMatch] = Field(..., description="Bugs detected from alert log")
    total_count: int = Field(..., description="Total bugs detected")
    summary: BugDetectionSummary = Field(..., description="Detection summary")


class VersionCheckResponse(BaseModel):
    """API response for version-specific bug check."""

    database_version: str = Field(..., description="Database version checked")
    bugs_affecting_version: List[Bug] = Field(
        ..., description="Bugs affecting this version"
    )
    total_count: int = Field(..., description="Total bugs found")
    critical_count: int = Field(..., description="Critical bugs count")
    recommendation: str = Field(..., description="Overall recommendation")

File: frontend/src/types/bug.ts
/**
 * Bug detection types matching backend Pydantic schemas
 */

/**
 * Evidence for bug detection
 */
export interface BugEvidence {
  execution_plan?: Record<string, any>;
  parameters?: Record<string, any>;
  query_characteristics?: Record<string, any>;
  wait_events?: Record<string, any>;
  sql_characteristics?: Record<string, any>;
  alert_log?: Record<string, any>;
}

/**
 * Oracle bug information
 */
export interface Bug {
  bug_number: string;
  title: string;
  category: string;
  severity: string;
  description: string;
  symptoms: string[];
  affected_versions: string[];
  fixed_versions: string[];
  workarounds: string[];
  remediation_sql?: string;
  my_oracle_support_url?: string;
}

/**
 * Detected bug match with confidence
 */
export interface BugMatch {
  bug: Bug;
  confidence: number;
  matched_patterns: string[];
  evidence: BugEvidence;
  sql_id?: string;
}

/**
 * Summary of bug detection results
 */
export interface BugDetectionSummary {
  total_bugs: number;
  by_severity: Record<string, number>;
  by_category: Record<string, number>;
  high_confidence_count: number;
  critical_count: number;
}

/**
 * Bug detection API response
 */
export interface BugDetectionResponse {
  sql_id?: string;
  detected_bugs: BugMatch[];
  summary: BugDetectionSummary;
  database_version?: string;
  detection_timestamp: string;
}

/**
 * Bug list API response
 */
export interface BugListResponse {
  bugs: Bug[];
  total_count: number;
  filters_applied?: Record<string, any>;
}

/**
 * Alert log bug detection response
 */
export interface AlertLogBugResponse {
  detected_bugs: BugMatch[];
  total_count: number;
  summary: BugDetectionSummary;
}

/**
 * Version-specific bug check response
 */
export interface VersionCheckResponse {
  database_version: string;
  bugs_affecting_version: Bug[];
  total_count: number;
  critical_count: number;
  recommendation: string;
}

/**
 * Bug detection filters
 */
export interface BugFilters {
  category?: string;
  severity?: string;
  version?: string;
}

/**
 * Bug severity levels
 */
export enum BugSeverity {
  CRITICAL = "critical",
  HIGH = "high",
  MEDIUM = "medium",
  LOW = "low",
}

/**
 * Bug categories
 */
export enum BugCategory {
  OPTIMIZER = "optimizer",
  EXECUTION = "execution",
  STATISTICS = "statistics",
  PARALLEL = "parallel",
  PARTITIONING = "partitioning",
  PARSING = "parsing",
  MEMORY = "memory",
  LOCKING = "locking",
  NETWORK = "network",
  OTHER = "other",
}

/**
 * Bug detection request parameters
 */
export interface BugDetectionParams {
  sql_id: string;
  database_version?: string;
}

/**
 * Version check request parameters
 */
export interface VersionCheckParams {
  database_version: string;
}
